lib/mongodb.js

```diff
@@ -537,7 +544,18 @@ MongoDB.prototype.coerceId = function(model, id, options) {
       }
     }
 
-    if (self.isObjectIDProperty(model, idProp, idValue, options)) {
+    const modelCtor = this._models[modelName];
+    if (shouldBeObjectID(idProp)) {
+      if (isObjectIDProperty(modelCtor, idProp, idValue, options)) {
+        idValue = ObjectID(idValue);
+      } else {
+        throw new Error(`${idValue} is not an ObjectID string`);
+      }
+    } else if (enableStrictObjectIDCoercion(modelCtor, options)) {
+      if (isObjectIDProperty(modelCtor, idProp, idValue, options)) {
+        idValue = ObjectID(idValue);
+      }
+    } else if (isObjectIDProperty(modelCtor, idProp, idValue, options)) {
```

Comment:
This code looks much better than before.

It makes me wonder though - do we still need this special rule in `coerceId` function? Would the coercion be already handled by `visitAllProperties`?

If there are cases that would not be handled by `visitAllProperties`, then please ensure there are tests covering them.